File: 1556-make-two-arrays-equal-by-reversing-subarrays/1556-make-two-arrays-equal-by-reversing-subarrays.cpp
class Solution {
public:
    bool canBeEqual(vector<int>& target, vector<int>& arr) {
        // if(target.size()==1 && target[0]==arr[0])
        //     return true;
        
        // sort(target.begin(), target.end());
        // sort(arr.begin(), arr.end());

        // for(int i=0; i<target.size(); i++)
        // {
        //     if(target[i]!=arr[i])
        //         return false;
        // }
        // return true;
        int n = target.size();
        map<int,int>mp;
        for(auto it:target)
            mp[it]++;

        for(int num:arr)
        {
            if(mp.find(num)==mp.end())
                return false;
            
            mp[num]--;
            if(mp[num]==0)
                mp.erase(num);
        }

        if(mp.size()==0)
            return true;
        
        return false;
    }
};

File: 1556-make-two-arrays-equal-by-reversing-subarrays/1556-make-two-arrays-equal-by-reversing-subarrays.py
class Solution:
    def canBeEqual(self, target: List[int], arr: List[int]) -> bool:
        # n = len(target)
        # arr.sort()
        # target.sort()

        # for i in range(n):
        #     if arr[i] != target[i]:
        #         return False
        
        # return True

        # using hashmap
    
        count_target = {}

        for num in target:
            if num in count_target:
                count_target[num] += 1
            else:
                count_target[num] = 1

        for num in arr:
            if num not in count_target:
                return False

            count_target[num] -= 1
            if count_target[num] == 0:
                del count_target[num]

        return len(count_target) == 0


        